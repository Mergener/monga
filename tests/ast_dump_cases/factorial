function factorial(x : int): int {
	var n : int;
	var f : float;
	var s : string;

	n = 1;
	while (x > 1) {
		n = n * x;
		x = x - 1;
	}

	return n;
}

function main(argc: int, argv: stringarray, f : float) {
	print(factorial(32));

	f = 4e+2;
}