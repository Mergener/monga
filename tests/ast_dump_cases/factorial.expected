<Module>
	<FunctionDefinition>
		<Name>factorial</Name>
		<ReturnType>int</ReturnType>
		<Parameters>
			<Parameter>
				<Name>x</Name>
				<Type>int</Type>
			</Parameter>
		</Parameters>
		<Block>
			<VariableDeclaration>
				<Name>n</Name>
				<Type>int</Type>
			</VariableDeclaration>
			<VariableDeclaration>
				<Name>f</Name>
				<Type>float</Type>
			</VariableDeclaration>
			<VariableDeclaration>
				<Name>s</Name>
				<Type>string</Type>
			</VariableDeclaration>
			<Assignment>
				<Var>n</Var>
				<IntConstant>1</IntConstant>
			</Assignment>
			<While>
				<Condition>
					<GreaterThan>
						<ExpVar>
							<Var>x</Var>
						</ExpVar>
						<IntConstant>1</IntConstant>
					</GreaterThan>
				</Condition>
				<Do>
					<Assignment>
						<Var>n</Var>
						<Multiplication>
							<OperandA>
								<ExpVar>
									<Var>n</Var>
								</ExpVar>
							</OperandA>
							<OperandB>
								<ExpVar>
									<Var>x</Var>
								</ExpVar>
							</OperandB>
						</Multiplication>
					</Assignment>
					<Assignment>
						<Var>x</Var>
						<Subtraction>
							<OperandA>
								<ExpVar>
									<Var>x</Var>
								</ExpVar>
							</OperandA>
							<OperandB>
								<IntConstant>1</IntConstant>
							</OperandB>
						</Subtraction>
					</Assignment>
				</Do>
			</While>
			<Return>
				<ExpVar>
					<Var>n</Var>
				</ExpVar>
			</Return>
		</Block>
	</FunctionDefinition>
	<FunctionDefinition>
		<Name>main</Name>
		<ReturnType>void</ReturnType>
		<Parameters>
			<Parameter>
				<Name>argc</Name>
				<Type>int</Type>
			</Parameter>
			<Parameter>
				<Name>argv</Name>
				<Type>stringarray</Type>
			</Parameter>
			<Parameter>
				<Name>f</Name>
				<Type>float</Type>
			</Parameter>
		</Parameters>
		<Block>
			<Call>
				<Callee>print</Callee>
				<Parameters>
					<Parameter>
						<Call>
							<Callee>factorial</Callee>
							<Parameters>
								<Parameter>
									<IntConstant>32</IntConstant>
								</Parameter>
							</Parameters>
						</Call>
					</Parameter>
				</Parameters>
			</Call>
			<Assignment>
				<Var>f</Var>
				<FloatConstant>400.000000</FloatConstant>
			</Assignment>
		</Block>
	</FunctionDefinition>
</Module>
